keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,823 @@
+
+1. Provide irreversable operation to kubectl users to convert management of an object's fields
+from client-side-apply to server-side-apply
+
+
+### Non-Goals
+
+1. Provide a server-side fix to this problem
+
+## Proposal
+
+Add a new CLI flag `--csa-manager`.
+
+When --server-side is used, fields owned by ---csa-manager will be converted to
+ownership of the ssa manager.
+
+If --csa-manager is not specified, then it is implied to be the same value as
+`--fieldmanager`
+
+
+```
+# Apply object with client-side-apply
+kubectl apply -f pod.yaml --fieldmanager mycsamanager
+```
+
+```
+# Apply object again with server-side-apply
+kubectl apply -f pod.yaml --fieldmanager=myssamanager --csa-manager=mycsamanager --server-side
+```
+
+With these flags, kubectl will migrate all fields owned by `mycsamanager` for client-side-apply
+to be now be owned by `myssamanager` for use with server-side-apply. This process
+is irreversible.
+
+### Inferred Defaults
+
+```
+# Apply object with client-side-apply
+kubectl apply -f pod.yaml
+```
+
+```
+# Apply object again with server-side-apply
+# Behind the scenes --csa-manager is inferred to be "kubectl-client-side-apply", 
+# and its fields are migrated to server-side-apply
+kubectl apply -f pod.yaml --server-side
+```
+
+These reasonable defaults are in line with what the kubernetes documentation says
+is the correct process for migrating to server-side apply, and users do not
+need to take extra steps.
+
+### Custom FieldManager
+
+If a custom `--fieldmanager` is specified in kubectl, that case behaves as expected as well
+
+```
+# Apply object with client-side-apply
+kubectl apply -f pod.yaml --fieldmanager mymanagername
+```
+
+```
+# Apply object again with server-side-apply
+# Behind the scenes --csa-manager is inferred to be "mymanagername"
+kubectl apply -f pod.yaml --fieldmanager=mymanagername --server-side
+```
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The bug this enhancement fixes is due to the fact that when `--server-side` is
+used today, it causes all managed fields owned by both the client-side-apply manager
+and the server-side-apply manager. So when the SSA manager drops a field, the CSA
+manager still owns it, and it is not deleted from the object:
+
+ManagedFIelds Before --server-side:
+
+
+After --server-side:
+
+
+After field is removed from SSA apply:
+
+
+### Approach
+
+Whenever `--server-side` is used with the `apply` subcommand, kubectl simply calls
+[`UpgradeManagedFields`](https://github.com/kubernetes/kubernetes/pull/111967/files#diff-4538195db8472f5237db69d0424dfd6fd7a7b0232f67f67dae52f57aea7b1af1R53) with the `--fieldmanager`
```

Comment:
What does it mean "calls" in this instance, is this an extra request made to the apiserver?